node/src/components/consensus/highway_core/active_validator/round_success_meter.rs

```diff
@@ -0,0 +1,100 @@
+use std::collections::VecDeque;
+
+use crate::types::Timestamp;
+
+const NUM_ROUNDS_TO_CONSIDER: usize = 40;
+const MAX_FAILED_ROUNDS: usize = 10;
+const ACCELERATION_PARAMETER: usize = 1000;
+
+pub(super) struct RoundSuccessMeter {
+    // store whether a particular round was successful
+    // index 0 is the last handled round, 1 is the second-to-last etc.
+    rounds: VecDeque<bool>,
+    round_exp: u8,
+    round_length: u64,
+    last_exponent_change_index: usize,
+    latest_handled_round_index: Option<usize>,
+}
+
+impl RoundSuccessMeter {
+    pub fn new(round_exp: u8, timestamp: Timestamp) -> Self {
+        let round_length = 1u64 << round_exp;
+        Self {
+            rounds: VecDeque::with_capacity(NUM_ROUNDS_TO_CONSIDER),
+            round_exp,
+            round_length,
+            last_exponent_change_index: (timestamp.millis() / round_length) as usize,
+            latest_handled_round_index: None,
+        }
+    }
+
+    pub fn change_exponent(&mut self, new_exp: u8, timestamp: Timestamp) {
+        self.rounds = VecDeque::with_capacity(NUM_ROUNDS_TO_CONSIDER);
+        self.round_exp = new_exp;
+        self.round_length = 1u64 << new_exp;
+        self.last_exponent_change_index = (timestamp.millis() / self.round_length) as usize;
+        self.latest_handled_round_index = None;
+    }
+
+    /// Registers the success or failure of finalizing a block within a round, and returns the new
+    /// round exponent based on whether we are keeping up with the blocks or not.
+    pub fn handle_finalized_block(
+        &mut self,
+        block_timestamp: Timestamp,
+        time_finalized: Timestamp,
+    ) -> u8 {
+        // in case the block timestamp isn't at the beginning of the round
```

Comment:
That can only happen if that block belongs to a round we're not even assigned to, can it?
I.e. if we have round length 32, and that block is at round 40, so its proposer must have round length 8.
I wonder whether we should ignore such blocks? I imagined success or failure of a round was determined by whether we finalized that round's own proposal in time, which by definition has a timestamp = round ID.

Reply:
I think we should be counting that. If for some reason we are on round exponent 6, while the whole network is at 5, and by chance we don't finalize a few of the blocks that were at the start of our rounds, but finalized all the other blocks - we wouldn't want to conclude that we're too fast and slow down even more, would we?

Reply:
Good point, yes. So as long as _any_ block was proposed and finalized within our round, we count that round as successful. :+1: